File: dialogs/src/main/java/com/rands/lightbulb/dialogs/picker_dialog_color/ColorPickerAdapter.java
package com.rands.lightbulb.dialogs.picker_dialog_color;

import android.content.Context;
import android.graphics.Color;
import android.graphics.drawable.Drawable;
import android.os.Parcel;
import android.os.Parcelable;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import com.rands.lightbulb.commons.utils.ResourceUtils;
import com.rands.lightbulb.dialogs.R;
import com.rands.lightbulb.recycleradapters.EasyAdapterConfiguration;
import com.rands.lightbulb.recycleradapters.EasyAdapterDataModel;
import com.rands.lightbulb.recycleradapters.EasyAdapterSelectableModes;
import com.rands.lightbulb.recycleradapters.EasyRecyclerAdapter;

import java.util.ArrayList;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.appcompat.widget.AppCompatImageButton;
import androidx.recyclerview.widget.RecyclerView;

import static com.rands.lightbulb.dialogs.picker_dialog_color.ColorPickerAdapter.*;

public class ColorPickerAdapter extends EasyRecyclerAdapter<ColorModel> {
    private final Context context;

    public ColorPickerAdapter(Context context, EasyAdapterSelectableModes selectableMode) {
        super(new EasyAdapterConfiguration<ColorModel>().withSelectMode(selectableMode));
        this.context = context;
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        AppCompatImageButton v = (AppCompatImageButton) LayoutInflater.from(parent.getContext())
                .inflate(R.layout.layout_color_item, parent, false);
        return new ColorVH(v);
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        ColorVH vh = (ColorVH) holder;
        ColorModel item = items.get(position);
        vh.setItem(item, position, this);
    }

    @Override
    public void onViewRecycled(@NonNull RecyclerView.ViewHolder holder) {
        super.onViewRecycled(holder);
        ColorVH vh = (ColorVH) holder;
        vh.release();
    }

    @Override
    public int getItemCount() {
        return items.size();
    }

    @Override
    public String getItemName(ColorModel item) {
        return item.getItemName();
    }

    @Override
    public void setCollection(List<ColorModel> collection) {
        super.setCollection(validatePendingCollection(collection));
    }

    @Override
    public void appendCollection(List<ColorModel> collection) {
        super.appendCollection(validatePendingCollection(collection));
    }

    public Drawable getColorDrawable(ColorModel colorModel) {
        if (colorModel == null)
            return null;
        Drawable colorDrawable = ResourceUtils.getDrawable(context, R.drawable.color_icon).mutate();
        colorDrawable.setTint(Color.parseColor(colorModel.colorHex));
        return colorDrawable;
    }

    private List<ColorModel> validatePendingCollection(List<ColorModel> inputCollection) {
        List<ColorModel> outputCollection = new ArrayList<>();
        for (ColorModel colorInput : inputCollection) {
            try {
                Color.parseColor(colorInput.colorHex);
                outputCollection.add(colorInput);
            } catch (Exception e) {
                //ignore
            }
        }
        return outputCollection;
    }

    public static class ColorVH extends RecyclerView.ViewHolder {
        protected AppCompatImageButton colorView;
        protected ColorModel item;

        ColorVH(AppCompatImageButton view) {
            super(view);
            this.colorView = (AppCompatImageButton) itemView;
        }

        public void setItem(ColorModel item, int position, ColorPickerAdapter adapter) {
            this.item = item;
            Context ctx = colorView.getContext();
            boolean isItemSelected = adapter.isItemSelected(position);
            int drawableId = isItemSelected ? R.drawable.color_selected_icon : R.drawable.color_icon;
            Drawable iconDrawable = ResourceUtils.getDrawable(ctx, drawableId).mutate();
            iconDrawable.setTint(Color.parseColor(item.colorHex));
            //colorView.setColorFilter(Color.parseColor(item.colorHex));
            colorView.setImageDrawable(iconDrawable);
            colorView.setOnClickListener(v -> {
                boolean prevState = adapter.isItemSelected(position);
                adapter.selectItemAt(position, !prevState);
            });
        }

        public void release() {
            colorView.setImageDrawable(null);
        }
    }

    public static class ColorModel extends EasyAdapterDataModel {
        private final String colorHex;
        private final String colorExternalName;

        public ColorModel(String iconName, String externalName) {
            super(false);
            this.colorExternalName = externalName;
            this.colorHex = iconName;
        }

        // Parcelling part
        public ColorModel(Parcel in) {
            super(in);
            this.colorHex = in.readString();
            this.colorExternalName = in.readString();
        }

        public static final Parcelable.Creator<ColorModel> CREATOR = new Parcelable.Creator<ColorModel>() {
            public ColorModel createFromParcel(Parcel in) {
                return new ColorModel(in);
            }

            public ColorModel[] newArray(int size) {
                return new ColorModel[size];
            }
        };

        @Override
        public void writeToParcel(Parcel dest, int i) {
            dest.writeString(colorHex);
            dest.writeString(colorExternalName);
        }

        @Override
        public int describeContents() {
            return 0;
        }

        @Override
        public String getItemName() {
            return colorExternalName;
        }

        public String getColorHex() {
            return colorHex;
        }

        public String getColorExternalName() {
            return colorExternalName;
        }
    }
}

File: lightbulb-dialogs/src/main/java/com/github/RooneyAndShadows/lightbulb/dialogs/picker_dialog_color/ColorPickerAdapter.kt
package com.github.rooneyandshadows.lightbulb.dialogs.picker_dialog_color

import android.content.Context
import android.graphics.Color
import android.graphics.drawable.Drawable
import android.os.Parcel
import android.os.Parcelable
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.appcompat.widget.AppCompatImageButton
import androidx.recyclerview.widget.RecyclerView
import com.github.rooneyandshadows.lightbulb.commons.utils.ResourceUtils
import com.github.rooneyandshadows.lightbulb.dialogs.R
import com.github.rooneyandshadows.lightbulb.dialogs.picker_dialog_color.ColorPickerAdapter.ColorModel
import com.github.rooneyandshadows.lightbulb.recycleradapters.abstraction.EasyAdapterDataModel
import com.github.rooneyandshadows.lightbulb.recycleradapters.abstraction.EasyAdapterSelectableModes
import com.github.rooneyandshadows.lightbulb.recycleradapters.abstraction.EasyRecyclerAdapter
import java.lang.Exception
import java.util.ArrayList

@Suppress("unused")
class ColorPickerAdapter(
    private val context: Context,
    selectableMode: EasyAdapterSelectableModes = EasyAdapterSelectableModes.SELECT_SINGLE
) : EasyRecyclerAdapter<ColorModel>(selectableMode) {
    @Override
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val v: AppCompatImageButton = LayoutInflater.from(parent.context)
            .inflate(R.layout.layout_color_item, parent, false) as AppCompatImageButton
        return ColorVH(v)
    }

    @Override
    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val vh = holder as ColorVH
        val item: ColorModel = getItem(position)!!
        vh.setItem(item, position, this)
    }

    @Override
    override fun onViewRecycled(holder: RecyclerView.ViewHolder) {
        super.onViewRecycled(holder)
        val vh = holder as ColorVH
        vh.release()
    }

    @Override
    override fun getItemName(item: ColorModel): String {
        return item.externalName
    }

    @Override
    override fun setCollection(collection: List<ColorModel>) {
        super.setCollection(validatePendingItems(collection))
    }

    @Override
    override fun appendCollection(collection: List<ColorModel>) {
        super.appendCollection(validatePendingItems(collection))
    }

    @Override
    override fun addItem(item: ColorModel) {
        validatePendingItem(item)?.apply {
            super.addItem(this)
        }
    }

    private fun validatePendingItems(collection: List<ColorModel>): List<ColorModel> {
        val outputCollection: MutableList<ColorModel> = ArrayList()
        for (colorInput in collection)
            validatePendingItem(colorInput)?.apply {
                outputCollection.add(this)
            }
        return outputCollection
    }

    private fun validatePendingItem(item: ColorModel): ColorModel? {
        return try {
            Color.parseColor(item.colorHex)
            item
        } catch (e: Exception) {
            null
        }
    }

    fun getColorDrawable(colorModel: ColorModel?): Drawable? {
        if (colorModel == null) return null
        val colorDrawable: Drawable = ResourceUtils.getDrawable(context, R.drawable.color_icon)!!.mutate()
        colorDrawable.setTint(Color.parseColor(colorModel.colorHex))
        return colorDrawable
    }

    class ColorVH internal constructor(view: AppCompatImageButton) : RecyclerView.ViewHolder(view) {
        private var colorView: AppCompatImageButton = itemView as AppCompatImageButton
        private var item: ColorModel? = null

        fun setItem(item: ColorModel, position: Int, adapter: ColorPickerAdapter) {
            this.item = item
            val ctx: Context = colorView.context
            val isItemSelected: Boolean = adapter.isItemSelected(position)
            val drawableId = if (isItemSelected) R.drawable.color_selected_icon else R.drawable.color_icon
            val iconDrawable: Drawable = ResourceUtils.getDrawable(ctx, drawableId)!!.mutate()
            iconDrawable.setTint(Color.parseColor(item.colorHex))
            //colorView.setColorFilter(Color.parseColor(item.colorHex));
            colorView.setImageDrawable(iconDrawable)
            colorView.setOnClickListener {
                val prevState: Boolean = adapter.isItemSelected(position)
                adapter.selectItemAt(position, !prevState)
            }
        }

        fun release() {
            colorView.setImageDrawable(null)
        }
    }

    @Suppress("unused")
    class ColorModel : EasyAdapterDataModel {
        val colorHex: String
        val externalName: String

        constructor(colorHex: String, externalName: String) : super() {
            this.colorHex = colorHex
            this.externalName = externalName
        }

        // Parcelling part
        constructor(parcel: Parcel) : super() {
            colorHex = parcel.readString()!!
            externalName = parcel.readString()!!
        }

        override fun writeToParcel(dest: Parcel, i: Int) {
            dest.writeString(colorHex)
            dest.writeString(externalName)
        }

        override val itemName: String
            get() = externalName

        override fun describeContents(): Int {
            return 0
        }

        companion object CREATOR : Parcelable.Creator<ColorModel> {
            override fun createFromParcel(parcel: Parcel): ColorModel {
                return ColorModel(parcel)
            }

            override fun newArray(size: Int): Array<ColorModel?> {
                return arrayOfNulls(size)
            }
        }
    }
}